flow/raster_cache.cc

```diff
@@ -175,25 +176,63 @@ std::unique_ptr<RasterCacheResult> RasterCache::RasterizeDisplayList(
 
 void RasterCache::Prepare(PrerollContext* context,
                           Layer* layer,
-                          const SkMatrix& ctm) {
-  RasterCacheKey cache_key(layer->unique_id(), RasterCacheKeyType::kLayer, ctm);
-  Entry& entry = cache_[cache_key];
+                          const SkMatrix& ctm,
+                          RasterCacheLayerStrategy strategy) {
+  auto cache_key_optional =
+      TryToMakeRasterCacheKeyForLayer(layer, strategy, ctm);
+  if (!cache_key_optional) {
+    return;
+  }
+  Entry& entry = cache_[cache_key_optional.value()];
   entry.access_count++;
   entry.used_this_frame = true;
   if (!entry.image) {
-    entry.image = RasterizeLayer(context, layer, ctm, checkerboard_images_);
+    entry.image =
+        RasterizeLayer(context, layer, strategy, ctm, checkerboard_images_);
+  }
+}
+
+std::optional<RasterCacheKey> RasterCache::TryToMakeRasterCacheKeyForLayer(
+    const Layer* layer,
+    RasterCacheLayerStrategy strategy,
+    const SkMatrix& ctm) const {
+  switch (strategy) {
+    case RasterCacheLayerStrategy::kLayer:
+      return RasterCacheKey(layer->unique_id(), RasterCacheKeyType::kLayer,
+                            ctm);
+    case RasterCacheLayerStrategy::kLayerChildren:
+      auto& children_layers =
+          static_cast<const ContainerLayer*>(layer)->layers();
```

Comment:
Do we ever try to cache anything other than a container layer? Perhaps we can change the signature to avoid these casts.

Reply:
It makes sense, but I was looking for a way that enforces the condition via the argument types. I can't say for sure that a non-container layer might want to cache itself, though, so the as_container_layer solution would only enforce the contract for layers using the kLayerChildren which is more targeted.